src/config.js

```diff
@@ -0,0 +1,1602 @@
+            return true
+          } else if (value.toLowerCase() === 'false') {
+            return false
+          } else {
+            throw new Error(`Validation failed at ${keyPath}, ${JSON.stringify(value)} must be a boolean or the string 'true' or 'false'`)
+          }
+        case 'boolean':
+          return value
+        default:
+          throw new Error(`Validation failed at ${keyPath}, ${JSON.stringify(value)} must be a boolean or the string 'true' or 'false'`)
+      }
+    }
+  },
+
+  'string': {
+    validate (keyPath, value, schema) {
+      if (typeof value !== 'string') {
+        throw new Error(`Validation failed at ${keyPath}, ${JSON.stringify(value)} must be a string`)
+      }
+      return value
+    },
+
+    validateMaximumLength (keyPath, value, schema) {
+      if ((typeof schema.maximumLength === 'number') && (value.length > schema.maximumLength)) {
+        return value.slice(0, schema.maximumLength)
+      } else {
+        return value
+      }
+    }
+  },
+
+  'null': {
+    // null sort of isnt supported. It will just unset in this case
+    coerce (keyPath, value, schema) {
+      if (![undefined, null].includes(value)) { throw new Error(`Validation failed at ${keyPath}, ${JSON.stringify(value)} must be null`) }
+      return value
+    }
+  },
+
+  'object': {
+    coerce (keyPath, value, schema) {
+      if (!isPlainObject(value)) { throw new Error(`Validation failed at ${keyPath}, ${JSON.stringify(value)} must be an object`) }
+      if (schema.properties == null) { return value }
+
+      let defaultChildSchema = null
+      let allowsAdditionalProperties = true
+      if (isPlainObject(schema.additionalProperties)) {
+        defaultChildSchema = schema.additionalProperties
+      }
+      if (schema.additionalProperties === false) {
+        allowsAdditionalProperties = false
+      }
+
+      const newValue = {}
+      for (let prop in value) {
+        const propValue = value[prop]
+        const childSchema = schema.properties[prop] != null ? schema.properties[prop] : defaultChildSchema
+        if (childSchema != null) {
+          try {
+            newValue[prop] = this.executeSchemaEnforcers(pushKeyPath(keyPath, prop), propValue, childSchema)
+          } catch (error) {
+            console.warn(`Error setting item in object: ${error.message}`)
+          }
+        } else if (allowsAdditionalProperties) {
+          // Just pass through un-schema'd values
+          newValue[prop] = propValue
+        } else {
+          console.warn(`Illegal object key: ${keyPath}.${prop}`)
+        }
+      }
+
+      return newValue
+    }
+  },
+
+  'array': {
+    coerce (keyPath, value, schema) {
+      if (!Array.isArray(value)) { throw new Error(`Validation failed at ${keyPath}, ${JSON.stringify(value)} must be an array`) }
+      const itemSchema = schema.items
+      if (itemSchema != null) {
+        const newValue = []
+        for (let item of Array.from(value)) {
```

Comment:
the check on 1513 verifies that this is an array so we don't need the `Array.from()`